box_error.go

```diff
@@ -0,0 +1,274 @@
+								return nil, err
+							}
+							if mapv, err = d.DecodeInterface(); err != nil {
+								return nil, err
+							}
+							errorStack[i].Fields[mapk] = mapv
+						}
+					default:
+						if err = d.Skip(); err != nil {
+							return nil, err
+						}
+					}
+				}
+
+				if i > 0 {
+					errorStack[i-1].Prev = &errorStack[i]
+				}
+			}
+		default:
+			if err = d.Skip(); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if len(errorStack) > 0 {
+		return &errorStack[0], nil
+	}
+
+	return nil, nil
+}
+
+func encodeBoxError(enc *encoder, boxError *BoxError) (err error) {
+	if err = enc.EncodeMapLen(1); err != nil {
+		return err
+	}
+	if err = encodeUint(enc, KeyErrorStack); err != nil {
+		return err
+	}
+
+	var stackDepth = boxError.Depth()
+	if err = enc.EncodeArrayLen(stackDepth); err != nil {
+		return err
+	}
+
+	for ; stackDepth > 0; stackDepth-- {
+		fieldsLen := len(boxError.Fields)
+
+		if fieldsLen > 0 {
+			if err = enc.EncodeMapLen(7); err != nil {
+				return err
+			}
+		} else {
+			if err = enc.EncodeMapLen(6); err != nil {
+				return err
+			}
+		}
+
+		if err = encodeUint(enc, KeyErrorType); err != nil {
+			return err
+		}
+		if err = enc.EncodeString(boxError.Type); err != nil {
+			return err
+		}
+
+		if err = encodeUint(enc, KeyErrorFile); err != nil {
+			return err
+		}
+		if err = enc.EncodeString(boxError.File); err != nil {
+			return err
+		}
+
+		if err = encodeUint(enc, KeyErrorLine); err != nil {
+			return err
+		}
+		if err = enc.EncodeUint32(boxError.Line); err != nil {
+			return err
+		}
+
+		if err = encodeUint(enc, KeyErrorMessage); err != nil {
+			return err
+		}
+		if err = enc.EncodeString(boxError.Msg); err != nil {
+			return err
+		}
+
+		if err = encodeUint(enc, KeyErrorErrno); err != nil {
+			return err
+		}
+		if err = enc.EncodeUint32(boxError.Errno); err != nil {
+			return err
+		}
+
+		if err = encodeUint(enc, KeyErrorErrcode); err != nil {
+			return err
+		}
+		if err = enc.EncodeUint32(boxError.Code); err != nil {
+			return err
+		}
+
+		if fieldsLen > 0 {
+			if err = encodeUint(enc, KeyErrorFields); err != nil {
+				return err
+			}
+
+			if err = enc.EncodeMapLen(fieldsLen); err != nil {
+				return err
+			}
+
+			for k, v := range boxError.Fields {
+				if err = enc.Encode(k); err != nil {
+					return err
+				}
+				if err = enc.Encode(v); err != nil {
+					return err
+				}
+			}
+		}
+
+		if stackDepth > 1 {
+			boxError = boxError.Prev
+		}
+	}
+
+	return nil
+}
+
+// MarshalMsgpack serializes the BoxError into a MessagePack representation.
+func (boxError *BoxError) MarshalMsgpack() (b []byte, err error) {
+	var buf smallWBuf
+
+	enc := newEncoder(&buf)
+	if err = encodeBoxError(enc, boxError); err != nil {
+		return b, err
+	}
+
+	return buf.b, nil
+}
+
+// UnmarshalMsgpack deserializes a BoxError value from a MessagePack
+// representation.
+func (boxError *BoxError) UnmarshalMsgpack(b []byte) error {
+	var buf smallBuf = smallBuf{b: b}
```

Comment:
It can be just a byte buffer. There is no need in smallBuf (which may be confusing).
```suggestion
	dec := newDecoder(bytes.NewBuffer(b))
```

Reply:
I had inherited this from common-purpose Marshal/Unmarshal, but it seems that plain buffer should be enough too. (But a little consistency sounds nice too.)

Reply:
Changed